### PR TITLE
fix(settlement): use block descriptors as source of truth for blocks to apply 

### LINKED ISSUE #1169: Make sure block descriptors is the source of truth when applying and validating
Rollback may change the block descriptors so we can't blindly count that it matches da blocks.

### PR DESCRIPTION
**Pull request by srene**
# PR Standards

## Opening a pull request should be able to meet the following requirements

--

PR naming convention: https://hackmd.io/@nZpxHZ0CT7O5ngTp0TP9mg/HJP_jrm7A

---

Close #1169 

<-- Briefly describe the content of this pull request -->

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: block/retriever.go [hunk 1]
```diff
@@ -6,38 +6,45 @@ import (
...
 
 	m.retrieverMu.Lock()
 	defer m.retrieverMu.Unlock()
 
 	var lastAppliedHeight float64
+	blockIndex := 0
 	for _, batch := range batchResp.Batches {
 		for i, block := range batch.Blocks {
+			// We dont apply a block if not included in the block descriptor (adds support for rollback)
+			if blockIndex >= len(slBatch.BlockDescriptors) {
```
Comment (omritoptix) [inline block/retriever.go:30]: wouldn't it be cleaner to check `block.Header.Height` against the heights height of `BlockDescriptor.Height` so in that case you don't need to add any index etc?

**Review by omritoptix — APPROVED** [review]
approved with small suggestion

File: block/retriever.go (hunk 1 above)
Comment (srene) [reply block/retriever.go:30]: i think the index is necessary anyway, if we want to keep compatiblity with  multiple batches in the same state update

Comment (srene) [reply block/retriever.go:30]: although not supported right now...

Comment (omritoptix) [reply block/retriever.go:30]: another problem I see now that we only increment the `blockIndex` at the end of the loop but we have `continue` in the middle which will make the block index incorrect..

Comment (srene) [reply block/retriever.go:30]: fixed

**Review by omritoptix — APPROVED** [review]

**Pull request merged**